File: OpenChess/static/js/game.js
const gameState = {
    board: [],
    currentTurn: "white",
    moveHistory: [],
    check: false,
    selectedSquare: null,
    legalMoves: [],
};

async function fetchState() {
    const response = await fetch("/api/state");
    const data = await response.json();
    applyState(data);
}

async function fetchLegalMoves(row, col) {
    const response = await fetch("/api/legal-moves", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ row, col }),
    });
    const data = await response.json();
    return data.legal_moves || [];
}

async function submitMove(fromRow, fromCol, toRow, toCol) {
    const response = await fetch("/api/move", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
            from_row: fromRow,
            from_col: fromCol,
            to_row: toRow,
            to_col: toCol,
        }),
    });
    return response.json();
}

async function resetGame() {
    await fetch("/api/reset", { method: "POST" });
    gameState.selectedSquare = null;
    gameState.legalMoves = [];
    await fetchState();
}

function applyState(data) {
    gameState.board = data.board;
    gameState.currentTurn = data.current_turn;
    gameState.moveHistory = data.move_history;
    gameState.check = data.check;
    draw();
    updateStatusPanel(gameState);
}

function draw() {
    renderBoard(gameState.board, {
        selectedSquare: gameState.selectedSquare,
        legalMoves: gameState.legalMoves,
        kingInCheckSquare: gameState.check ? findKingSquare(gameState.board, gameState.currentTurn) : null,
        onSquareClick: handleSquareClick,
    });
}

function findKingSquare(board, color) {
    for (let row = 0; row < 8; row += 1) {
        for (let col = 0; col < 8; col += 1) {
            const piece = board[row][col];
            if (piece && piece.color === color && piece.type === "king") {
                return { row, col };
            }
        }
    }
    return null;
}

async function handleSquareClick(row, col) {
    const clickedPiece = gameState.board[row][col];

    if (gameState.selectedSquare) {
        const isLegalTarget = gameState.legalMoves.some((move) => move[0] === row && move[1] === col);

        if (isLegalTarget) {
            const result = await submitMove(
                gameState.selectedSquare.row,
                gameState.selectedSquare.col,
                row,
                col
            );
            gameState.selectedSquare = null;
            gameState.legalMoves = [];
            if (result.success) {
                await fetchState();
            } else {
                draw();
            }
            return;
        }

        if (clickedPiece && clickedPiece.color === gameState.currentTurn) {
            await selectSquare(row, col);
            return;
        }

        gameState.selectedSquare = null;
        gameState.legalMoves = [];
        draw();
        return;
    }

    if (clickedPiece && clickedPiece.color === gameState.currentTurn) {
        await selectSquare(row, col);
    }
}

async function selectSquare(row, col) {
    gameState.selectedSquare = { row, col };
    gameState.legalMoves = await fetchLegalMoves(row, col);
    draw();
}

document.addEventListener("DOMContentLoaded", () => {
    fetchState();
    document.getElementById("reset-button").addEventListener("click", resetGame);
});

File: chess/game.py
from chess.board import Board
from chess.rules import get_legal_moves, is_king_in_check
from chess.constants import WHITE, BLACK


class Game:
    def __init__(self):
        self.board = Board()
        self.current_turn = WHITE
        self.move_history = []

    def get_legal_moves_for_square(self, row, col):
        piece = self.board.get_piece(row, col)
        if piece is None or piece.color != self.current_turn:
            return []
        return get_legal_moves(self.board, row, col)

    def make_move(self, from_row, from_col, to_row, to_col):
        piece = self.board.get_piece(from_row, from_col)
        if piece is None or piece.color != self.current_turn:
            return {"success": False, "error": "No piece of the current player on that square"}

        legal_moves = get_legal_moves(self.board, from_row, from_col)
        if (to_row, to_col) not in legal_moves:
            return {"success": False, "error": "That move is not legal"}

        captured_piece = self.board.move_piece(from_row, from_col, to_row, to_col)

        self.move_history.append({
            "piece": piece.piece_type,
            "color": piece.color,
            "from": [from_row, from_col],
            "to": [to_row, to_col],
            "captured": captured_piece.piece_type if captured_piece else None,
        })

        self.current_turn = BLACK if self.current_turn == WHITE else WHITE

        return {
            "success": True,
            "captured": captured_piece.piece_type if captured_piece else None,
            "current_turn": self.current_turn,
            "check": is_king_in_check(self.board, self.current_turn),
        }

    def get_state(self):
        return {
            "board": self.board.to_dict(),
            "current_turn": self.current_turn,
            "move_history": self.move_history,
            "check": is_king_in_check(self.board, self.current_turn),
        }

    def reset(self):
        self.__init__()
